File: src/com/silverkeytech/android_rivers/DownloadService.java
package com.silverkeytech.android_rivers;

import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.*;
import android.util.Log;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;

public class DownloadService extends Service {

    private static final String TAG = DownloadService.class.getSimpleName();

    SharedPreferences preferences;

    private static final String DOCUMENT_VIEW_STATE_PREFERENCES = "DjvuDocumentViewState";

    private Looper mServiceLooper;
    private ServiceHandler mServiceHandler;
    private NotificationManager mNM;
    String downloadUrl;
    String fileName;
    int targetFileSize;

    public static boolean serviceState = false;

    // Handler that receives messages from the thread
    private final class ServiceHandler extends Handler {
        public ServiceHandler(Looper looper) {
            super(looper);
        }

        @Override
        public void handleMessage(Message msg) {
            downloadFile();
            showNotification("Downloading", "VVS");
            stopSelf(msg.arg1);
        }
    }


    @Override
    public void onCreate() {
        serviceState = true;
        mNM = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
        HandlerThread thread = new HandlerThread("ServiceStartArguments", 1);
        thread.start();

        // Get the HandlerThread's Looper and use it for our Handler
        mServiceLooper = thread.getLooper();
        mServiceHandler = new ServiceHandler(mServiceLooper);

    }

    public void downloadFile() {
        downloadFile(this.downloadUrl, fileName);
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.d(TAG, "SERVICE-ONCOMMAND - onStartCommand");

        Bundle extra = intent.getExtras();
        if (extra != null) {
            String downloadUrl = extra.getString("downloadUrl");
            Log.d(TAG, "DOWNLOAD URL FROM INTENT URL " + downloadUrl);
            this.downloadUrl = downloadUrl;

            String filename = extra.getString("filename");
            Log.d(TAG, "Target name " + filename);
            this.fileName = filename;

            int targetFile = extra.getInt("contentLength");
            Log.d(TAG,  "Target content size " + targetFile);
            this.targetFileSize = targetFile;
        }

        Message msg = mServiceHandler.obtainMessage();
        msg.arg1 = startId;
        mServiceHandler.sendMessage(msg);

        // If we get killed, after returning from here, restart
        return START_STICKY;
    }


    @Override
    public void onDestroy() {
        Log.d(TAG, "SERVICE-DESTROY - DESTORY");
        serviceState = false;
        //Toast.makeText(this, "service done", Toast.LENGTH_SHORT).show();
    }


    @Override
    public IBinder onBind(Intent intent) {
        // We don't provide binding, so return null
        return null;
    }

    void showNotification(String message, String title) {
        // In this sample, we'll use the same text for the ticker and the expanded notification
        CharSequence text = message;

        // Set the icon, scrolling text and timestamp
        Notification notification = new Notification(android.R.drawable.gallery_thumb, "vvs",
                System.currentTimeMillis());

        notification.flags |= Notification.FLAG_AUTO_CANCEL;
        Intent intent = new Intent(this, MainActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        //The PendingIntent to launch our activity if the user selects this notification
        PendingIntent contentIntent = PendingIntent.getActivity(this.getBaseContext(), 0,
                intent, PendingIntent.FLAG_CANCEL_CURRENT);

        // Set the info for the views that show in the notification panel.
        notification.setLatestEventInfo(this, title,
                text, contentIntent);
        // Send the notification.
        // We use a layout id because it is a unique number.  We use it later to cancel.
        mNM.notify(R.string.app_name, notification);
    }

    public void downloadFile(String fileURL, String fileName) {

        StatFs stat_fs = new StatFs(Environment.getExternalStorageDirectory().getPath());
        double avail_sd_space = (double) stat_fs.getAvailableBlocks() * (double) stat_fs.getBlockSize();
        //double GB_Available = (avail_sd_space / 1073741824);
        double MB_Available = (avail_sd_space / 10485783);
        //System.out.println("Available MB : " + MB_Available);

        Log.d(TAG, "available MB " + MB_Available);

        try {
            Log.d(TAG, "Before checking externalStorageDirectory");
            File root = new File(Environment.getExternalStorageDirectory() + "/dody");
            if (root.exists() && root.isDirectory()) {

            } else {
                root.mkdir();
            }

            Log.d(TAG, "CURRENT PATH " + root.getPath());
            Log.d(TAG, "Opening " + fileURL);


            URL u = new URL(fileURL);
            HttpURLConnection c = (HttpURLConnection) u.openConnection();
            c.setRequestMethod("GET");
            c.setDoOutput(true);
            Log.d(TAG, "Before connecting");
            c.connect();

            int fileSize = c.getContentLength() / 1048576;
            Log.d(TAG, "FILESIZE " + fileSize);

            if (MB_Available <= fileSize) {
                this.showNotification("NOTIFICATION NO MEMORY", "NOTIFICATION ERROR");
                c.disconnect();
                return;
            }

            Log.d(TAG, "Before opening output stream");
            FileOutputStream f = new FileOutputStream(new File(root.getPath(), fileName));

            Log.d(TAG, "Before getting the remote input stream");
            InputStream in = c.getInputStream();


            byte[] buffer = new byte[1024];
            int len1 = 0;
            int loop = 0;
            while ((len1 = in.read(buffer)) > 0) {
                f.write(buffer, 0, len1);
                loop += 1024;

                Float percentage = ((float) loop / targetFileSize) * 100;

                Log.d(TAG, "Progressing " + percentage + "% download " + loop + " with target " + this.targetFileSize);

            }
            f.close();

            Log.d(TAG, "Download complete");

        } catch (Exception e) {
            Log.d(TAG, "Downloader There is an exception : " + e.getMessage());
        }
    }
}

File: src/com/silverkeytech/android_rivers/DownloadService.kt
package com.silverkeytech.android_rivers

import android.app.IntentService
import android.content.Intent
import android.util.Log
import com.github.kevinsawicki.http.HttpRequest
import com.github.kevinsawicki.http.HttpRequest.HttpRequestException
import android.os.Environment
import java.io.File
import android.app.Activity
import android.os.Messenger
import android.os.Message
import android.os.RemoteException
import android.support.v4.app.NotificationCompat
import android.app.PendingIntent
import android.widget.RemoteViews
import android.content.Context
import android.app.NotificationManager
import android.app.Notification
import java.util.Random
import java.net.URL
import java.io.BufferedInputStream
import java.io.FileOutputStream


public class DownloadService() : IntentService("DownloadService"){
    class object{
        public val PARAM_DOWNLOAD_URL : String= "downloadUrl"
        public val PARAM_DOWNLOAD_TITLE : String = "downloadTitle"
        public val PARAM_DOWNLOAD_LOCATION_PATH : String = "downloadLocationPath"

        public val TAG: String = javaClass<DownloadService>().getSimpleName()!!
    }

    var targetUrl : String? = null
    var targetTitle : String? = null

    fun prepareNotification(inferredName : String, filePath : String) : Notification{
        var notificationIntent = Intent(Intent.ACTION_MAIN)
        notificationIntent.setClass(getApplicationContext(), javaClass<MainActivity>())

        notificationIntent.putExtra(PARAM_DOWNLOAD_LOCATION_PATH, filePath)

        var contentIntent = PendingIntent.getActivity(this, 0, notificationIntent, PendingIntent.FLAG_CANCEL_CURRENT)

        var notification = NotificationCompat.Builder(this)
                .setTicker("Podcast starts downloading")
        ?.setSmallIcon(android.R.drawable.gallery_thumb)
        ?.setProgress(100,10, true)
        ?.setWhen(System.currentTimeMillis())
        ?.setContentIntent(contentIntent)
        ?.build()

        notification!!.icon = android.R.drawable.star_big_on

        notification!!.contentView = RemoteViews(getApplicationContext()!!.getPackageName(), R.layout.download_progress)

        notification!!.contentView!!.setImageViewResource(R.id.download_progress_status_icon, android.R.drawable.btn_star);
        notification!!.contentView!!.setProgressBar(R.id.download_progress_status_progress, 100, 0, false)
        notification!!.contentView!!.setTextViewText(R.id.download_progress_status_text, "Downloading $targetTitle")

        return notification!!
    }

    protected override fun onHandleIntent(p0: Intent?) {
        targetUrl = p0!!.getStringExtra(PARAM_DOWNLOAD_URL)
        targetTitle = p0!!.getStringExtra(PARAM_DOWNLOAD_TITLE)

        Log.d(TAG, "onHandleIntent with ${targetUrl}")

        var inferredName = getFileNameFromUri(targetUrl!!)

        var result = Activity.RESULT_CANCELED
        var filename : String = ""

        var notification : Notification? = null

        var notificationId = Random().nextLong().toInt()
        var notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        try{
            var url = URL(targetUrl)
            var connection = url.openConnection()!!
            connection.connect()

            var fileLength : Int = connection.getContentLength()

            Log.d(TAG, "File length $fileLength")

            if (inferredName == null)
                inferredName = generateThrowawayName() + ".mp3"

            var directory = Environment.getExternalStorageDirectory()!!.getPath() + "/" + Environment.DIRECTORY_PODCASTS
            filename  = directory + "/" + inferredName

            notification = prepareNotification(inferredName!!, filename)

            Log.d(TAG, "Podcast to be stored at ${filename}")

            notificationManager.notify(notificationId, notification)

            var input = BufferedInputStream(url.openStream()!!)
            var output = FileOutputStream(filename)

            var data =  ByteArray(1024)
            Log.d(TAG, "Length of byte array ${data.size}")
            var total : Long = 0
            var progress : Int = 0

            var count : Int = input.read(data)

            var oldProgress = 0

            while (count != -1){
                total += count
                oldProgress = progress
                progress = (total.toInt() * 100) div fileLength
                Log.d(TAG, "Download progress ($total * 100) / $fileLength $progress")

                if (progress != oldProgress){
                    notification!!.contentView!!.setProgressBar(R.id.download_progress_status_progress, 100, progress, false)
                    notificationManager.notify(notificationId, notification)
                }

                output.write(data, 0, count)

                count = input.read(data)
            }

            output.flush()
            output.close()
            input.close()

            notification!!.contentView!!.setTextViewText(R.id.download_progress_status_text, "File successfully download to $filename")
            notificationManager.notify(notificationId, notification)

            result = Activity.RESULT_OK
        }
        catch(e : HttpRequestException){
            Log.d(TAG, "Exception happend at attempt to download ${e.getMessage()}")
            notification!!.contentView!!.setTextViewText(R.id.download_progress_status_text, "File $inferredName download cancelled")
            notificationManager.notify(notificationId, notification)
        }

        var extras = p0.getExtras()
        if (extras != null){
            var messenger = extras!!.get(Params.MESSENGER) as android.os.Messenger
            var msg = Message.obtain()!!
            msg.arg1 = result
            msg.obj = filename
            try{
                messenger.send(msg)
            }
            catch(e : RemoteException){
                Log.d(TAG, "Have problem when try to send a message ")
            }
        }
    }

    public override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        Log.d(TAG, "OnStartCommand  with ${targetUrl}")

        return super<IntentService>.onStartCommand(intent, flags, startId)
    }

    public override fun onCreate() {
        super<IntentService>.onCreate()
        Log.d(TAG, "Service created  with ${targetUrl}")
    }

    public override fun onStart(intent: Intent?, startId: Int) {
        super<IntentService>.onStart(intent, startId)
        Log.d(TAG, "Service started  with ${targetUrl}")
    }

    public override fun onDestroy() {
        super<IntentService>.onDestroy()
        Log.d(TAG, "Service destroyed  with ${targetUrl}")
    }
}
